fw/http_frame.c

```diff
@@ -2235,50 +2266,54 @@ do {									\
 			 * zeroed if client close this stream.
 			 */
 			BUG_ON(!tfw_h2_stream_is_active(stream));
-			SCHED_REMOVE_NOT_EXCLUSIVE_STREAM(sched, stream);
 		} else if (ctx->error && tfw_h2_stream_is_active(ctx->error)) {
 			stream = ctx->error;
-			SCHED_REMOVE_NOT_EXCLUSIVE_STREAM(sched, stream);
-			error_was_sent = true;
+			/*
+			 * If we send error response we stop to send any data
+			 * from other streams, so we either sent all error
+			 * response or blocked by window size.
+			*/
+			stop = true;
 		} else {
-			stream = tfw_h2_sched_stream_dequeue(sched, &parent);
+			stream = tfw_h2_sched_get_most_prio_stream(sched);
 		}
 
 		/*
 		 * If root scheduler is active we always can find
 		 * active stream.
 		 */
 		BUG_ON(!stream);
-		r = tfw_h2_stream_xmit_process(sk, ctx, stream, !parent,
-					       &snd_wnd);
-
-		/* We don't recalculate deficits of exclusive streams. */
-		if (parent) {
-			deficit = tfw_h2_stream_recalc_deficit(stream);
-			tfw_h2_sched_stream_enqueue(sched, stream, parent,
-						    deficit);
-		} else if (!tfw_h2_stream_is_active(stream)) {
+		stream_is_exclusive = tfw_h2_stream_is_exclusive(stream);
+		r = tfw_h2_stream_xmit_process(sk, ctx, stream,
+					       stream_is_exclusive,
+					       &snd_wnd, &stop);
+
+		if (!tfw_h2_stream_is_active(stream)) {
 			tfw_h2_sched_deactivate_stream(sched, stream);
-			/*
-			 * Remove exclusive stream after sending all pending
-			 * data.
-			 */
-			if (!stream->xmit.skb_head)
-				tfw_h2_stream_clean(ctx, stream);
+			if (!stream->xmit.skb_head) {
+				/*
+				 * Remove exclusive stream after sending all
+				 * pending data.
+				 */
+				if (stream_is_exclusive) {
+					tfw_h2_stream_clean(ctx, stream);
+				} else {
+					TfwStreamSchedEntry *parent =
+						stream->sched->parent;
+
+					tfw_h2_stream_sched_remove(sched,
+								   stream);
+					tfw_h2_sched_stream_enqueue(sched,
+								    stream,
+								    parent);
```

Comment:
This code is frequently called and I think we can aggregate both the functions to get some advantage, mostly in the propagation functions. This pair of functions is also called from `tfw_h2_stream_sched_move_child()` - we can just rewrite the function and use it here.

Please also make the function names consistent: use `sched` and `stream` in the same order